File: partiql-ast/src/main/java/org/partiql/ast/v1/graph/GraphPart.java
package org.partiql.ast.v1.graph;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.partiql.ast.v1.AstNode;
import org.partiql.ast.v1.AstVisitor;
import org.partiql.ast.v1.expr.Expr;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
public abstract class GraphPart extends AstNode {
    /**
     * TODO docs, equals, hashcode
     */
    @Builder(builderClassName = "Builder")
    @EqualsAndHashCode(callSuper = false)
    public static class Node extends GraphPart {
        @Nullable
        public final Expr prefilter;

        @Nullable
        public final String variable;

        @Nullable
        public final GraphLabel label;

        public Node(@Nullable Expr prefilter, @Nullable String variable, @Nullable GraphLabel label) {
            this.prefilter = prefilter;
            this.variable = variable;
            this.label = label;
        }

        @Override
        @NotNull
        public Collection<AstNode> children() {
            List<AstNode> kids = new ArrayList<>();
            if (prefilter != null) {
                kids.add(prefilter);
            }
            if (label != null) {
                kids.add(label);
            }
            return kids;
        }

        @Override
        public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
            return visitor.visitGraphPartNode(this, ctx);
        }
    }

    /**
     * TODO docs, equals, hashcode
     */
    @Builder(builderClassName = "Builder")
    @EqualsAndHashCode(callSuper = false)
    public static class Edge extends GraphPart {
        @NotNull
        public final GraphDirection direction;

        @Nullable
        public final GraphQuantifier quantifier;

        @Nullable
        public final Expr prefilter;

        @Nullable
        public final String variable;

        @Nullable
        public final GraphLabel label;

        public Edge(@NotNull GraphDirection direction, @Nullable GraphQuantifier quantifier,
        @Nullable Expr prefilter, @Nullable String variable, @Nullable GraphLabel label) {
        this.direction = direction;
        this.quantifier = quantifier;
        this.prefilter = prefilter;
        this.variable = variable;
        this.label = label;
    }

        @Override
        @NotNull
        public Collection<AstNode> children() {
            List<AstNode> kids = new ArrayList<>();
            if (quantifier != null) {
                kids.add(quantifier);
            }
            if (prefilter != null) {
                kids.add(prefilter);
            }
            if (label != null) {
                kids.add(label);
            }
            return kids;
        }

        @Override
        public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
            return visitor.visitGraphPartEdge(this, ctx);
        }
    }

    /**
     * TODO docs, equals, hashcode
     */
    @Builder(builderClassName = "Builder")
    @EqualsAndHashCode(callSuper = false)
    public static class Pattern extends GraphPart {
        @NotNull
        public final GraphPattern pattern;

        public Pattern(@NotNull GraphPattern pattern) {
            this.pattern = pattern;
        }

        @Override
        @NotNull
        public Collection<AstNode> children() {
            List<AstNode> kids = new ArrayList<>();
            kids.add(pattern);
            return kids;
        }

        @Override
        public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
            return visitor.visitGraphPartPattern(this, ctx);
        }
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/graph/GraphPart.kt
package org.partiql.ast.v1.graph

import org.partiql.ast.v1.AstNode
import org.partiql.ast.v1.AstVisitor
import org.partiql.ast.v1.expr.Expr

/**
 * TODO docs, equals, hashcode
 */
public abstract class GraphPart : AstNode() {
    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = when (this) {
        is Node -> visitor.visitGraphPartNode(this, ctx)
        is Edge -> visitor.visitGraphPartEdge(this, ctx)
        is Pattern -> visitor.visitGraphPartPattern(this, ctx)
        else -> throw NotImplementedError()
    }

    /**
     * TODO docs, equals, hashcode
     */
    public class Node(
        @JvmField
        public var prefilter: Expr?,
        @JvmField
        public var variable: String?,
        @JvmField
        public var label: GraphLabel?,
    ) : GraphPart() {
        public override fun children(): Collection<AstNode> {
            val kids = mutableListOf<AstNode?>()
            prefilter?.let { kids.add(it) }
            label?.let { kids.add(it) }
            return kids.filterNotNull()
        }

        public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
            visitor.visitGraphPartNode(this, ctx)
    }

    /**
     * TODO docs, equals, hashcode
     */
    public class Edge(
        @JvmField
        public var direction: GraphDirection,
        @JvmField
        public var quantifier: GraphQuantifier?,
        @JvmField
        public var prefilter: Expr?,
        @JvmField
        public var variable: String?,
        @JvmField
        public var label: GraphLabel?,
    ) : GraphPart() {
        public override fun children(): Collection<AstNode> {
            val kids = mutableListOf<AstNode?>()
            quantifier?.let { kids.add(it) }
            prefilter?.let { kids.add(it) }
            label?.let { kids.add(it) }
            return kids.filterNotNull()
        }

        public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
            visitor.visitGraphPartEdge(this, ctx)
    }

    /**
     * TODO docs, equals, hashcode
     */
    public class Pattern(
        @JvmField
        public var pattern: GraphPattern,
    ) : GraphPart() {
        public override fun children(): Collection<AstNode> {
            val kids = mutableListOf<AstNode?>()
            kids.add(pattern)
            return kids.filterNotNull()
        }

        public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
            visitor.visitGraphPartPattern(this, ctx)
    }
}
